### ISSUE TITLE
false positive for type-hinted Enum class instance variable

### ISSUE THREAD

**Issue by huwcbjones** [pylint-dev/pylint#7402]
### Bug description

As per https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#classes, you can optionally declare instance variables in the class body. For normal classes, this is fine, but for Enums pylint raises a false positive

Define an enum with a type-hinted instance variable and you get an invalid name! 😞 
```py
class MyEnum(str, Enum):
    attr: int

    VALUE = "a", 1

    def __new__(cls, value, attr=None):
        obj = str.__new__(cls, value)
        obj._value_ = value
        obj.attr = attr
        return obj
```

### Command used

```shell
pylint --disable=all --enable=invalid-name mwe.py
```
(`pylint mwe.py` also works, I just wanted to filter out the irrelavent messages)

### Pylint output

```shell
************* Module mwe
mwe.py:5:4: C0103: Class constant name "attr" doesn't conform to UPPER_CASE naming style (invalid-name)

--------------------------------------------------------------------
Your code has been rated at 8.89/10
```

### Expected behavior

No error

### Pylint version

```shell
pylint 2.12.2/2.15.0
astroid 2.9.3/2.12.5
Python 3.9.1 (default, Mar 15 2021, 18:22:18)
[Clang 12.0.0 (clang-1200.0.32.27)]
```

### OS / Environment

macOS

**Comment by huwcbjones** [comment]
From my quick peruse of the code, I believe that pylint will treat all Enum class variables as class constants, even if no value is assigned (the variable is for a type hint).
https://github.com/PyCQA/pylint/blob/880095c7266134a218e5f97075d1adeaa5b34b79/pylint/checkers/base/name_checker/checker.py#L459-L464

**Comment by mbyrnepr2** [comment]
Thanks for reporting. I think this is expected since Enum members are expected to be constants. Quoting from the docs:

_Because Enums are used to represent constants we recommend using UPPER_CASE names for enum members, and will be using that style in our examples._

https://docs.python.org/3/library/enum.html

**Comment by Pierre-Sassoulas** [comment]
I'd also say it's working as intended.

**Comment by andrei-korshikov** [comment]
@mbyrnepr2 not quite so. `VALUE` (from the example above) is enum member, it is constant, and it should be upper case. But `attr` is not enum member, it is **the attribute of enum member,** so it is just ordinary attribute of instance, it's **not class constant.**

As reference example, look at the [Planet](https://docs.python.org/3/howto/enum.html#planet) from Python docs. They don't write `self.MASS`, but `self.mass`. So for type hinting we have to write `mass: int` with the lower case in the class `Planet (Enum)` body.

And here is my own example (that is why I've dug out this issue):
```python
from enum import Enum

class Color(Enum):
    """Represents colors as (red, green, blue) tuples."""

    YELLOW      = 250, 250,   0
    KHAKI       = 250, 250, 125
    MAGENTA     = 250,   0, 250
    VIOLET      = 250, 125, 250
    CYAN        =   0, 250, 250
    AQUAMARINE  = 125, 250, 250

    # Class constant name "red" doesn't conform to UPPER_CASE naming style (invalid-name)
    # Class constant name "green" doesn't conform to UPPER_CASE naming style (invalid-name)
    # Class constant name "blue" doesn't conform to UPPER_CASE naming style (invalid-name)
    red: int
    green: int
    blue: int

    def __init__(self, red: int, green: int, blue: int) -> None:
        self.red   = red
        self.green = green
        self.blue  = blue

    @property
    def as_hex(self) -> str:
        """Get hex 'abcdef' representation for a color."""
        return f'{self.red:0{2}x}{self.green:0{2}x}{self.blue:0{2}x}'
```

@Pierre-Sassoulas I disagree with you:) @huwcbjones clearly showed to us that Pylint sees no difference between enum member and its instance variables. Is it possible to reopen the issue?

**Comment by mbyrnepr2** [comment]
Thank you @andrei-korshikov 👍 Completely agree with you.
I am working on fixing it 🎉

**Comment by DanielNoord** [comment]
@mbyrnepr2 In which PR has this been fixed? Might be good to link here.

**Comment by mbyrnepr2** [comment]
Two changes for this fix. A Pylint PR: https://github.com/pylint-dev/pylint/pull/8905 and an astroid PR: https://github.com/pylint-dev/astroid/pull/2263
